text/3334-niche.md

```diff
@@ -0,0 +1,348 @@
+unsigned integers. The unsigned integers may use any integer base
+representation (decimal, hex, binary, octal), but must not have a type suffix.
+The unsigned integers are interpreted as the bit patterns in memory
+corresponding to the representation of the non-ZST field. For instance, a
+struct with a float field could specify one or more NaN values as a niche using
+the integer representation of those values.
+
+The attribute `#[niche]` may only appear on a struct declaration. The struct
+must contain exactly one field of a non-zero-sized type (non-ZST). The struct
+may contain zero or more ZST fields, such as `PhantomData`. (Note that
+`#[non_exhaustive]` types do not count as ZSTs for this purpose, even if they
+*currently* contain no fields with non-zero sizes.)
+
+Declaring a niche on any item other than a struct declaration results in an
+error.
+
+Declaring a niche on a struct containing more or less than one non-zero-sized
+field results in an error.
+
+Declaring multiple `niche` attributes on a single item, or multiple key-value
+pairs within a single `niche` attribute, results in an error.
+
+Declaring a niche on a struct that has any generic parameters affecting the
+non-zero-sized field results in an error.
+
+Declaring a range niche with an empty range (e.g. `0..0`) results in a
+warn-by-default lint. As with many lints, this lint should be automatically
+suppressed for code expanded from a macro.
+
+Declaring a range niche with an invalid range (e.g. `5..0`) results in an
+error.
+
+Declaring a niche using a negative value or a negative range endpoint results
+in an error. The representation of negative values depends on the size of the
+type, and the compiler may not have that information at the time it handles
+attributes such as `niche`. The text of the error should suggest the
+appropriate two's-complement unsigned equivalent to use. The compiler may
+support this in the future.
+
+Declaring a range niche with an open start (`..3`) results in an error, for
+forwards-compatibility with support for negative values.
+
+Declaring a niche using a non-literal value (e.g. `usize::MAX`) results in an
+error. Constants can use compile-time evaluation, and compile-time evaluation
+does not occur early enough for attributes such as niche declarations.
+
+If a type `T` contains multiple niche values (e.g. `#[niche(range = 8..16)]`),
+the compiler does not guarantee any particular usage of those niche values in
+the representation of types containing `T`. In particular, the
+compiler does not commit to making use of all the invalid values of the niche,
+even if it otherwise could have.
+
+However, multiple instances of the same identical type (e.g. `Option<T>` and
+`Option<T>`) will use an identical representation (whether the type contains a
+niche or not). This permits a round-trip between such a value and a byte
+representation.
+
+If a type `T` contains niches and uses `repr(C)` or `repr(transparent)`, the
```

Comment:
From the perspective of "Define them as narrowly as possible", could we just say instead "this is only allowed with `repr(transparent)`"?  That would mean we don't need to repeat the transparent rules ("The struct must contain exactly one field of a non-zero-sized type (non-ZST)." from line 93, and similar) in their definition, for example.

Is there any situations in which `repr(C)` is better than `repr(transparent)` for one of these types?  Is there any situation in which one wouldn't want `repr(transparent)` on it?